ZENKO-1999 - cosmos scheduler refactor

ssalaues opened the pull request:
**What does this PR do, and why do we need it?**
Refactors the cosmos scheduler code making it more legible, extensible, and improves logging. 

Additionally:
- it is more deterministic with reading the configuration overlays
- Fixes remaining possible racy and deadlock scenarios
- More easily allows for future locations to be added

Commit:
refactor: comments, better logs, and legibility

- Moves instanciating of concurrency-unsafe objects into the goroutines
- Makes it more easily extensible to watch other location types

Commit:
bugfix: fixes race between bucket watch and overlays

Commit:
ft: go docs

Commit:
ft: only watch for configuration overlay

- Specifically only watch for any new config overlays and updates to those keys
instead of watching for all updates on the entire PENSIEVE collection. This
reduces the amount of updates for a single Orbit update from 4 down to 1.

bert-e commented:
# Hello ssalaues,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/zenko/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Unsupported issue type



The Jira issue ZENKO-1999 is of type Refactoring and cannot be used to reference a branch.


I can only merge the following issue types into development branches:

Jira issue type  |  corresponding branch prefix
-----------------|------------------
`Bug` | `bugfix`
`Design` | `design`
`Documentation` | `documentation`
`Feature` | `feature`
`Improvement` | `improvement`
`Story` | `feature`

rachedbenmustapha commented on cosmos/scheduler/cmd/scheduler.go:
```diff
@@ -32,33 +34,50 @@ type Scheduler struct {
+		err := s.MongodbClient.Ping(ctx, readpref.Nearest())
 		if err != nil {
 			log.Println("MongoDB healthcheck failed:", err)
 			http.Error(w, "Internal error, could not contact MongoDB", 500)
```
In idiomatic go, there would be a `return` here, leaving the normal code path un-indented.

rachedbenmustapha commented on cosmos/scheduler/cmd/scheduler.go:
```diff
@@ -76,65 +95,73 @@ func (s *Scheduler) healthCheckServer() {
 	secret := s.getIngestionSecret()
 	if secret == nil {
 		log.Println("creating ingestion secret")
 		err = s.ingestionSecret(accessKey, secretKey, false)
```
The name `ingestionSecret` is a bit confusing, it looks like a getter. Something with write or save or set in the name would be less surprising.

rachedbenmustapha commented on cosmos/scheduler/cmd/scheduler.go:
```diff
@@ -76,65 +95,73 @@ func (s *Scheduler) healthCheckServer() {
-			return err
+			log.Println("error creating ingestion secret", err)
+		} else {
+			log.Println("ingestion secret created successfully")
```
The non-error flow is usually un-indented

rachedbenmustapha commented on cosmos/scheduler/main.go:
```diff
@@ -53,8 +53,13 @@ func init() {
+	if len(mongoOptions.Hosts) == 1 {
+		mongoOptions.SetReadPreference(readpref.Nearest())
+	} else {
+		mongoOptions.SetReadPreference(readpref.Secondary())
```
Wouldn't we always want the primary?

ssalaues replied:
My rationale behind this was because this program will never need to write to MongoDB hence the preference on the secondary (although secondaries will be unavailable singleton deployments). This way there is more available connections to the primary for services that do require it.

ssalaues replied:
Switched this to the Primary.

rachedbenmustapha commented on cosmos/scheduler/cmd/scheduler.go:
```diff
@@ -76,65 +95,73 @@ func (s *Scheduler) healthCheckServer() {
 	ch := make(chan BucketTransaction)
+	defer close(ch)
 	go func() {
+		collection := s.MongodbClient.Database("metadata").Collection("__metastore")
```
I know this is not new in this PR, but something to take care of later: the database name should be configurable (sandbox deployment uses `MONGODB_DATABASE`), for cases where zenko instances are co-hosted.

ssalaues replied:
Yeah that's a good point. I'll add this in.

rachedbenmustapha commented on cosmos/scheduler/cmd/scheduler.go:
```diff
@@ -76,65 +95,73 @@ func (s *Scheduler) healthCheckServer() {
+// for the locationType. A new bucket created in the locationType will trigger an applyChanges
+// for the bucket. If a true is sent on the bool channel then the function will return closing all
+// open connections.
+func (s *Scheduler) watchBucketUpdates(locationType string, quit chan bool) error {
```
Instead of having a special quit channel, a context can be passed as the first argument, and in the `select` loop, use `case <-ctx.Done():`.

The argument context can then be used as-is (with or without a cancellation wrapper, depending on the need) in nested context-aware function calls.

This allows a top-level `Context` cancellation to trickle down, and makes standard easy to read function signatures (once one gets used to them).

ssalaues replied:
Yeah this is one I kept changing to try to get rid of the quit channel but couldn't come up with a good solution. I think this is great suggestion and will make these changes.

rachedbenmustapha commented on cosmos/scheduler/cmd/scheduler.go:
```diff
@@ -32,33 +34,50 @@ type Scheduler struct {
 		ctx, cancel := context.WithTimeout(req.Context(), 5*time.Second)
 		defer cancel()
-		err := s.MongodbClient.Ping(ctx, readpref.Secondary())
+		err := s.MongodbClient.Ping(ctx, readpref.Nearest())
```
The health check should probably use the same read preference as the standard code path, to catch the same kind of problems. For example the health check can be happy reading from a secondary, while the application itself is failing reads on the primary.

rachedbenmustapha commented on cosmos/scheduler/cmd/scheduler.go:
```diff
@@ -32,33 +34,50 @@ type Scheduler struct {
+		// blocks until an overlay update is recieved on the overlayUpdates channel
+		<- overlayUpdates
 		log.Println("received an overlay update")
 		quit <- true
```
Why quit here? It seems that this is doing "watch, receive, quit" in a loop, which has the potential to miss events because of the asynchronicity involved.

ssalaues replied:
Agreed, I think I can more easily rework this now with your previous comments.

rachedbenmustapha reviewed (commented):
first pass

Commit:
ft: configurable database name

ssalaues commented on cosmos/scheduler/cmd/scheduler.go:
```diff
@@ -32,39 +35,57 @@ type Scheduler struct {
+		<- overlayUpdates
+		cancelNfs()
 		log.Println("received an overlay update")
-		quit <- true
```
TODO: the loop is still essentially watch, update, quit. Will still be reworking this.

Commit:
refactor: remove watch, receive, quit loop

This commit refactors the main Run loop to prevent missed events by allowing
both watchOverlayUpdates and watchBucketUpdates to run independantly of each
other. These changes allow for more extensibility as well in the future when
adding additonaly cosmos locations.

ssalaues commented on cosmos/scheduler/cmd/scheduler.go:
```diff
@@ -32,39 +39,58 @@ type Scheduler struct {
-		<-overlayUpdates
-		log.Println("received an overlay update")
-		quit <- true
+		select {
```
@RachedBenMustapha I replaced the "watch, receive, quit" loop here with a select loop where the goroutines are no longer interdependent on each other or kill/recreated on every update. 

It's added as a separate commit to make reviewing a bit easier:
https://github.com/scality/Zenko/pull/792/commits/554002e92739824489f8725fee78a4e0df8fd926

ssalaues commented on cosmos/scheduler/cmd/scheduler.go:
```diff
@@ -17,12 +19,17 @@ import (
 )
+const (
+	nfsLocation = "location-nfs-mount-v1"
+	scalityRing = "location-ring-s3-v1"
```
Was temporarily using `scalityRing` as an explicit check to return early but no longer needed. This can be deleted.

rachedbenmustapha reviewed (approved):
Looks more solid than it was, good improvement.

Suggestions for next steps:
- integrate `go vet` and `go fmt` checks in the CI (or even better, something like https://golangci.com/), there are anomalies that I didn't mention, that these tools would bring up automatically
- Use a single top-level context that can be passed down to all goroutines, and that can be used as global cancelation for `SIGTERM` for example
- Unless mongodb's watches always return a first result when set up, this only seems to be catching mongodb changes happening while the process is running, not while it's being restarted, upgraded, crashed, etc.

ssalaues commented:
@RachedBenMustapha thanks for the feedback and great suggestions for steps moving forward. The last one especially has been annoying not only for me but for the user experience as well.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZENKO-1999 contains:


* *None*


Considering where you are trying to merge, I expected to find:


* `1.1.2`


Please check the `Fix Version/s` of ZENKO-1999, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @ssalaues

* @vrancurel

* @nicolast

* @tmacro

ssalaues commented:
Rebased and added the version bump to Zenko.

ssalaues commented:
@bert-e approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/1.1`



The following branches will **NOT** be impacted:

* `development/0.9`
* `development/1.0`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.1`



The following branches have **NOT** changed:

* `development/0.9`
* `development/1.0`



Please check the status of the associated issue ZENKO-1999.

Goodbye ssalaues.
